### PR TITLE
Add support for nodeless local Gardener setup

### LINKED ISSUE #1812: Local dev setup without k8s cluster
**What would you like to be added**:
It would be nice to have an option for local development without the need for a local k8s cluster (like docker-for-desktop or kind).
@rfranzke @zanetworker proposed to use standalone` kube-apiserver` and `etcd` containers for that.

**Why is this needed**:
This would make the local setup more lightweight and easier wrt to choice of k8s versions and being independent of docker-for-desktop updates (which might break your local gardener setup).

### PR DESCRIPTION
**Pull request by zanetworker**
**What this PR does / why we need it**:
This PR adds support for a `nodeless` local Gardener setup.  With this setup, Kubernetes on docker is not required, the only requirement is `Docker`. All the components for this cluster will run as containers. 

For more information, please have a look at the documentation (`hack/local-garden/README.md`).

**Which issue(s) this PR fixes**:
Fixes #1812

**Special notes for your reviewer**:

**Release note**:

```improvement user
Developers can now make use of a `nodeless` local gardener setup. Simply run `make local-garden-up` and get a nodeless Garden cluster. 
```

### CODE REVIEW

File: hack/dev-setup-register-gardener [hunk 1]
```diff
@@ -19,16 +19,36 @@ source $(dirname "${0}")/common
 IP_ROUTE=$(ip route get 1)
 IP_ADDRESS=$(echo ${IP_ROUTE#*src} | awk '{print $1}')
 
+APISERVER_SERVICE_NAME="gardener-apiserver"
+APISERVER_ENDPOINT_NAME="gardener-apiserver"
+APISERVER_SERVICE_PORT=443
+APISERVICE_PORT_STRING=""
+
+APISERVER_EXTERNALNAME=local-apiserver
```
Comment (timebertt) [inline hack/dev-setup-register-gardener]:
```suggestion
APISERVER_EXTERNALNAME=kubernetes.localhost
```

Can we use something like `kubernetes.localhost` or `kubernetes.local` as `/etc/hosts` entry to make it clear, that this is a local entry?

File: hack/dev-setup-register-gardener [hunk 2]
```diff
@@ -19,16 +19,36 @@ source $(dirname "${0}")/common
...
 GARDEN_APISERVICE_NAME="v1beta1.garden.sapcloud.io"
 SETTINGS_APISERVICE_NAME="v1alpha1.settings.gardener.cloud"
 
-APISERVER_SERVICE_NAME="gardener-apiserver"
-APISERVER_ENDPOINT_NAME="gardener-apiserver"
 CONTROLLER_MANAGER_SERVICE_NAME="gardener-controller-manager"
 CONTROLLER_MANAGER_ENDPOINT_NAME="gardener-controller-manager"
 
+if [[ $(k8s_env) == "$NODELESS" ]]; then
+  APISERVICE_PORT_STRING="    port: $APISERVER_SERVICE_PORT"
```
Comment (timebertt) [inline hack/dev-setup-register-gardener]: IMHO, this is kind of cumbersome. Why is this needed?

File: hack/dev-setup-register-gardener [hunk 3]
```diff
@@ -128,6 +164,7 @@ spec:
   service:
     name: gardener-apiserver
     namespace: garden
+$APISERVICE_PORT_STRING
```
Comment (timebertt) [inline hack/dev-setup-register-gardener]:
```suggestion
    port: $APISERVER_SERVICE_PORT
```

File: Makefile [hunk 4]
```diff
@@ -34,6 +34,28 @@ dev-setup:
...
+	@./hack/local-garden/run-kube-controller-manager
+
+	# This etcd will be used to storge gardener resources (e.g., seeds, shoots)
+	@./hack/local-garden/run-gardener-etcd
+
+	# Applying proxy RBAC for the extension controller
+	# After this step, you can start using the cluster at hack/local-garden/kubeconfigs/admin.conf
+	@./hack/local-garden/apply-proxy-rbac
+
+	# Now you need to run ./hack/dev-setup-register-gardener
```
Comment (timebertt) [inline Makefile]: I guess you still have to create the garden Namespace yourself, right?
Can you automate that also or add a line for it?

File: hack/dev-setup-register-gardener (hunk 3 above)
Comment (timebertt) [reply hack/dev-setup-register-gardener]: Can you explain, why this is needed?
I didn't quite get it...

**Review by timebertt — COMMENTED** [review]
Very nice setup! :)
I will test it in my setup next week.

File: hack/dev-setup-register-gardener (hunk 2 above)
Comment (zanetworker) [reply hack/dev-setup-register-gardener]: Cross compatibility. Docker for desktop (i.e., k8s 1.14) does not support the port parameter for the apiservice as well.

File: hack/dev-setup-register-gardener (hunk 3 above)
Comment (zanetworker) [reply hack/dev-setup-register-gardener]: if it is docker, the port won't be set (docker does not support port since it runs on max 1.14)

File: Makefile (hunk 4 above)
Comment (zanetworker) [reply Makefile]: Can do, but the rest of the steps are left up for the developer.

Comment (timebertt) [reply Makefile]: Thanks, this is easier now 👍

File: hack/dev-setup-register-gardener (hunk 3 above)
Comment (timebertt) [reply hack/dev-setup-register-gardener]: Alright. Got it.

File: Makefile [hunk 5]
```diff
@@ -34,6 +34,28 @@ dev-setup:
 dev-setup-extensions:
 	@./hack/dev-setup-extensions
 
+.PHONY: dev-setup-local-garden
+dev-setup-local-garden:
```
Comment (timebertt) [inline Makefile]: It would be nice to also have a target to stop this nodeless cluster and remove the Docker containers and network.
WDYT of renaming this target to `local-cluster-up` and adding a second target `local-cluster-down` for removing the containers and the network?

**Review by timebertt — COMMENTED** [review]
Thanks for addressing my comments.

I also tried using this setup, but I always get logs like these from the Gardener API server:
```
E0127 11:40:35.439570   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.441397   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.532406   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.532552   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.532571   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.676281   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.676338   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
E0127 11:40:35.676947   56876 authentication.go:89] Unable to authenticate the request due to an error: x509: certificate signed by unknown authority
```

Not sure, what I am doing wrong.
Maybe you can help me out here.

**Comment by zanetworker** [comment]
@tim-ebert I don't think you are doing anything wrong, the gapi does not establish TLS with getcd and thus you are seeing this message, you should however be able to get gardener resources, is that the case?

File: Makefile (hunk 5 above)
Comment (zanetworker) [reply Makefile]: thanks for the suggestion, renamed to local-garden-up/down.

File: hack/dev-setup-register-gardener [hunk 6]
```diff
@@ -19,15 +19,35 @@ source $(dirname "${0}")/common
 IP_ROUTE=$(ip route get 1)
 IP_ADDRESS=$(echo ${IP_ROUTE#*src} | awk '{print $1}')
 
+APISERVER_SERVICE_NAME="gardener-apiserver"
+APISERVER_ENDPOINT_NAME="gardener-apiserver"
+APISERVER_SERVICE_PORT=443
+APISERVICE_PORT_STRING=""
+
+APISERVER_EXTERNALNAME=kubernetes.localhost
```
Comment (timuthy) [inline hack/dev-setup-register-gardener]:
```suggestion
APISERVER_EXTERNALNAME=gardener.localhost
```

File: hack/dev-setup-register-gardener [hunk 7]
```diff
@@ -19,15 +19,35 @@ source $(dirname "${0}")/common
...
+  APISERVICE_PORT_STRING="    port: $APISERVER_SERVICE_PORT"
+
+  echo "Adding API_SERVER address mapping to /etc/hosts!"
+
+  # remove entry from /etc/hosts
+  sudo sed -i".bak" "/$APISERVER_EXTERNALNAME/d" /etc/hosts
+
+  # then add it to make sure that addition is idempotent
+  LINE="$IP_ADDRESS\t$APISERVER_EXTERNALNAME"
+  sudo -- sh -c -e "echo '$LINE' >> /etc/hosts";
```
Comment (timuthy) [inline hack/dev-setup-register-gardener]: Instead of adjusting the `hosts` file of the host, can we rather adjust only the one inside the `kube-apiserver` container?

File: hack/local-garden/run-kube-etcd [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+MOUNTS="-v $SCRIPTPATH/certificates/certs:/certs -v $SCRIPTPATH/certificates/keys:/keys -v $ETCD_DATA_DIR:/etcd-data"
+
+
+echo "Starting gardener-dev kube-etcd cluster..!"
+docker run -d --name etcd -l $LABEL --network gardener-dev --rm  $MOUNTS $IMAGE etcd --name $CLUSTERNAME  \
+ --listen-client-urls "https://0.0.0.0:${ETCDPORT}" \
+ --advertise-client-urls "https://0.0.0.0:${ETCDPORT}" \
+ --listen-peer-urls "https://0.0.0.0:${PORT}" \
+ --trusted-ca-file "certs/ca.crt" \
+ --peer-trusted-ca-file "certs/ca.crt" \
```
Comment (timuthy) [inline hack/local-garden/run-kube-etcd:30]: Do we really need all that `peer` settings for such a simple setup?

File: hack/local-garden/kubeconfigs/certs/controller-manager.crt [hunk 9]
```diff
@@ -0,0 +1,24 @@
+-----BEGIN CERTIFICATE-----
```
Comment (timuthy) [inline hack/local-garden/kubeconfigs/certs/controller-manager.crt]: Can you in general create certificates which are valid for a longer period? At the moment it's only a year.

**Review by timuthy — CHANGES_REQUESTED** [review]
Nice PR @zanetworker :) I left you a few comments.

File: hack/local-garden/kubeconfigs/certs/controller-manager.crt (hunk 9 above)
Comment (zanetworker) [reply hack/local-garden/kubeconfigs/certs/controller-manager.crt]: Good point, will do that.

File: hack/local-garden/run-kube-etcd (hunk 8 above)
Comment (zanetworker) [reply hack/local-garden/run-kube-etcd:30]: now not really, but in the future with multiple `etcds` it can be useful to have (as for now, it won't be really used).

**Comment by timuthy** [comment]
@zanetworker I remember you saying that you wanted to automate the certificate handling or generation for such a local setup. Is this sill something you're pursuing?

**Comment by zanetworker** [comment]
@timuthy its already integrated in this PR :)

**Comment by timuthy** [comment]
Oh, sorry I missed that one :) Nice !

**Review by timuthy — APPROVED** [review]

File: hack/local-garden/cleanup [hunk 10]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+LABEL=${1:-local-garden}
+
+docker rm -f $(docker ps -f "label=$LABEL" -aq) 2&> /dev/null
```
Comment (timuthy) [inline hack/local-garden/cleanup]:
```suggestion
docker rm -f $(docker ps -f "label=$LABEL" -aq) 2&> /dev/null || true
```
see: https://unix.stackexchange.com/questions/212725/gnu-make-does-not-ignore-failed-command

File: hack/local-garden/cleanup [hunk 11]
```diff
@@ -0,0 +1,6 @@
...
+docker network rm $(docker network ls -qf "label=$LABEL")  2&> /dev/null
```
Comment (timuthy) [inline hack/local-garden/cleanup]:
```suggestion
docker network rm $(docker network ls -qf "label=$LABEL")  2&> /dev/null || true
```

**Review by timuthy — CHANGES_REQUESTED** [review]
Just a small thing that got my attention. LGTM otherwise.

**Review by timuthy — APPROVED** [review]
